blog-post.md

```diff
@@ -0,0 +1,292 @@
+that has become a barrier to its own evolution. Foundational features like 
+generics (aka parametric polymorphism) remain out of reach, as integrating them 
+into the current architecture is both difficult and dangerously prone to 
+introduce critical flaws.
+
+Aiming to break this impasse, the Argot collective decided to embark on a 
+fundamental rebuild, which is named Core Solidity: a project to define a new, 
+foundational version of the language designed for security, clarity, and 
+future growth from the ground up.
+
+
+## What is Core Solidity?
+
+The design of Core Solidity took inspiration from well established concepts from 
+functional programming, with Haskell serving as a primary influence, while 
+incorporating rigorous safety guarantees inspired by modern systems programming 
+languages such as Rust. Although this architectural direction represents a 
+significant evolution, it keeps the domain-specific requirements of smart 
+contract development, preserving the established conventions and practical 
+utility of Classic Solidity where appropriate.
+
+The language's theoretical foundation comprises several formally-specified 
+features that provide enhanced type safety and expressivity. In what follows, we 
+provide an overview of such features. However, we warn the reader that the current 
+Core Solidity prototype uses a syntax which is similar, but not identical, to 
+Classic Solidity. Since our initial concern is to develop the new language type 
+system and its semantics, its surface syntax would probably change in the near 
+future. All presented examples work in the current prototype.
+
+### Generics and Type Classes
+
+Core Solidity will introduce two new exciting abstraction mechanisms: generics and 
+type classes.
+
+Generics enable parametric polymorphism through type parameters, which allows the  
+implementation of algorithms and data structures that operate uniformly across 
+over all types. As an example, we could define a polymorphic `identity` function:
+
+``` 
+forall a . function identity(x : a) -> a {
+   return x;
+}
+```
+
+While generic functions are interesting, most interesting operations are not defined 
+at all types. Overloading allows the definition of code which can operate in distinct 
+ways at different types. Type classes are the standard way of combining overloading and 
+parametric polymorphism (generics) in a systematic manner. A type class definition 
+declares the class name, its arguments and member functions type signature. As an 
+example, let's consider the task of defining addition over different types:
+```
+function a . class a : Sum {
+  function sum (x : a, y : a) -> a;
+}
+```
+Implementations for member functions for different types are provided by instance 
+definitions. As an example, let's consider the implementation of `Sum` for 
+`word` type which, internally, uses Yul assembly to perform addition. 
+```
+instance word : Sum {
+  function sum (x : word, y : word) -> word {
```

Comment:
Can we make the example a bit more practical by doing a sum of two arbitrary `uintX` types? Or a concatenation of two arbitrary `bytesN` types?

The current examples are ok as illustrations of the concept but we should also use them to make people a bit more excited about Core Solidity by showing how easy it is to do stuff that is really tedious and repetitive in Classic Solidity.

Reply:
The first comment does not seem to have been addressed, but you also did not say if you disagree about it.

Reply:
We discussed this extensively on the chat. Here I'll just note that we decided to replace the example with `console.log()`.